### PR TITLE
Implement VERNAC EXTEND in coqpp

### PR DESCRIPTION
**Pull request by ppedrot**
Quite straightforward. Only one file has been ported, as all other users of this macro also declare `ARGUMENT EXTEND` extensions.

### CODE REVIEW

File: dev/doc/changes.md
Comment (gares) [inline dev/doc/changes.md]: can't you add this "context" in your preprocessor? I don't see why forcing people to return an unbound variable can be helpful (today and even tomorrow when st will be bound for real).

Comment (gares) [reply dev/doc/changes.md]: Said otherwise, I'd have `VERNAC....` that is the one that expects the st to be returned and `LEGACY VERNAC ...` that is supposed to die and adds the `let () = ... in st` wrapper. So that when you remove LEGACY you get a sensible type error.

Comment (ppedrot) [reply dev/doc/changes.md]: I didn't really understand @ejgallego's plan on this feature. The current situation does not make sense at all, because the returned state is never taken into account by the vernac interpretation. Adding a LEGACY marker would ease the transition, but the default case will require fixing anyways as it is broken...

Comment (ejgallego) [reply dev/doc/changes.md]: In order to handle locality we had to change the extension type, so I opted for a quite general one instead of having to define many. That being said feel free to refine with attributes or something.

Comment (gares) [reply dev/doc/changes.md]: In other words the current code is `LEGACY` and there is no infrastructure to remove the `LEGACY` (given that the result of functional code is discarded).

In any case, I prefer the code written between `{}` to make sense (in the old imperative model or in the new functional and only half baked mode). This PR makes the code just look silly, so please hide the silly context in the macro (with or without LEGACY, this is up to you).

Comment (ejgallego) [reply dev/doc/changes.md]: I wouldn't worry too much about this, but if you want, make the return type an option type, so the top level interpreter can know what to do (`None` would mean legacy)

Anyways, I feel that problems with the ongoing migration to pure state handling shouldn't delay this PR too much.

**Comment by herbelin** [comment]
I'm a bit confused. According to the CODEOWNERS file, @maximedenes should be assignee, but he is not even  in the list of reviewers.

**Comment by ppedrot** [comment]
Finally I made the port behave as the non-FUNCTIONAL extension, because the functional API is never going to make it in 8.9. So, for me this is now ready to go.

**Review by gares — APPROVED** [review]

**Review by aspiwack — APPROVED** [review]
My review is nominally requested, but the changes to my bits are trivial. So, on my end, nothing to report.

File: plugins/derive/g_derive.mlg
Comment (ppedrot) [reply plugins/derive/g_derive.mlg]: This allows to parse OCaml code without having to write an OCaml parser, using the same kind of trick as in e.g. melt. We have to wrap quotes inside delimiters that are well-parenthesized in the object language, so that it is just a matter of keeping a count of how many delimiters are opened so far (and handling a few constructs like strings and comments).

Comment (herbelin) [reply plugins/derive/g_derive.mlg]: I guess the question was about why moving from square brackets to curly brackets to delimit OCaml code. I guess that the reason is by consistency with the newly introduced curly brackets in `mlg` files. OK, you could ask why not to use square brackets in `mlg` files by consistency with `VERNAC EXTEND`. Then, I would guess that this is by consistency with `ocamlyacc`. The hysterical paths followed by syntax are often tortuous...

Comment (ppedrot) [reply plugins/derive/g_derive.mlg]: You have to change at least one of the uses of brackets into something else, because you cannot use it both for quotes and as a token. I went for braces for OCaml code because this was minimal in changes regarding camlp5-style grammar extension.

Comment (aspiwack) [reply plugins/derive/g_derive.mlg]: It was about both. And the answers are perfectly clear, thanks :slightly_smiling_face:

**Comment by maximedenes** [comment]
@ppedrot please fix https://github.com/coq/coq/issues/8405 and https://github.com/coq/coq/issues/8018 before porting more files to coqpp, otherwise it is going to make everybody's life more painful. Thanks!

**Comment by ejgallego** [comment]
So does this PR preserve a variant of the functional vernac extend?

We are very very close to actually use it for real.

**Comment by ppedrot** [comment]
No, but if this lands in master I can adapt the PR for that.

**Comment by ejgallego** [comment]
So who's gonna be the assignee here? Should I?

**Comment by maximedenes** [comment]
> No, but if this lands in master I can adapt the PR for that.

Is this ok with you, @ejgallego ?

**Comment by ejgallego** [comment]
> Is this ok with you, @ejgallego ?

It is OK to me.

**Comment by maximedenes** [comment]
> What I exactly mean is that the `.ml` files produced by coqp5 can use `coq.gramlib` instead of `camlp5.gramlib`, right?

@ejgallego do you want this change before merging this PR?

**Comment by ejgallego** [comment]
> @ejgallego do you want this change before merging this PR?

I think no change regarding the vendoring of camlp5 is needed here. This is [at the moment] purely a linking issue. Only problem I see with this PR is that a lot of code is not tested/used, but I think we can trust @ppedrot to implement the missing bits.

**Comment by maximedenes** [comment]
> I think no change regarding the vendoring of camlp5 is needed here. This is [at the moment] purely a linking issue. Only problem I see with this PR is that a lot of code is not tested/used, but I think we can trust @ppedrot to implement the missing bits.

Ok, thanks, so merging.

**Pull request merged**
